runtime/lua/vim/lsp/structures/diagnostic.lua

```diff
@@ -0,0 +1,230 @@
+  end
+  return diagnostics_by_line
+end
+
+local _diagnostic_counts = function(diagnostics)
+  if not diagnostics then return end
+
+  local counts = {}
+  for _, diagnostic in pairs(diagnostics) do
+    if diagnostic.severity then
+      if counts[diagnostic.severity] == nil then
+        counts[diagnostic.severity] = 0
+      end
+
+      counts[diagnostic.severity] = counts[diagnostic.severity] + 1
+    end
+  end
+
+  return counts
+end
+
+local set_diagnostic_cache = function(diagnostics, bufnr)
+  diagnostic_cache[bufnr] = diagnostics
+  diagnostic_cache_lines[bufnr] = _diagnostic_lines(diagnostics)
+  diagnostic_cache_counts[bufnr] = _diagnostic_counts(diagnostics)
+end
+
+local clear_diagnostic_cache = function(bufnr)
+  diagnostic_cache[bufnr] = nil
+  diagnostic_cache_lines[bufnr] = nil
+  diagnostic_cache_counts[bufnr] = nil
+end
+
+Diagnostic.save_buf_diagnostics = function(diagnostics, bufnr)
+  validate {
+    diagnostics = {diagnostics, 't', true};
+    bufnr = {bufnr, 'n', true};
+  }
+
+  if not diagnostics then return end
+  bufnr = bufnr == 0 and api.nvim_get_current_buf() or bufnr
+
+  if not diagnostic_cache[bufnr] then
+    -- Clean up our data when the buffer unloads.
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(b)
+        clear_diagnostic_cache(b)
+      end
+    })
+  end
+
+  set_diagnostic_cache(diagnostics, bufnr)
+end
+
+Diagnostic.set_virtual_text = function(diagnostics, bufnr, diagnostic_ns)
+  if not diagnostics then
+    return
+  end
+
+  diagnostic_ns = diagnostic_ns or default_diagnostic_ns
+
+  local buffer_line_diagnostics = diagnostic_cache_lines[bufnr] or _diagnostic_lines(diagnostics)
+
+  for line, line_diagnostics in pairs(buffer_line_diagnostics) do
+    local virt_texts = {}
+    for i = 1, #line_diagnostics - 1 do
+      table.insert(virt_texts, {"■", severity_highlights[line_diagnostics[i].severity]})
+    end
+    local last = line_diagnostics[#line_diagnostics]
+    -- TODO(ashkan) use first line instead of subbing 2 spaces?
+    table.insert(virt_texts, {"■ "..last.message:gsub("\r", ""):gsub("\n", "  "), severity_highlights[last.severity]})
+    api.nvim_buf_set_virtual_text(bufnr, diagnostic_ns, line, virt_texts, {})
+  end
+end
+
+-- TODO: I don't like that this function is "special" in that doesn't take any diagnostics as first arg.
+-- TODO: Rename to something like Diagnostic.get_saved_counts or similar
+Diagnostic.get_counts = function(bufnr, kind)
+  return (diagnostic_cache_counts[bufnr] or {})[protocol.DiagnosticSeverity[kind]]
+end
+
+Diagnostic.set_signs = function(diagnostics, bufnr, sign_ns)
+  sign_ns = sign_ns or default_sign_ns
+  for _, diagnostic in ipairs(diagnostics) do
+    vim.fn.sign_place(
+      0,
+      sign_ns,
+      diagnostic_severity_map[diagnostic.severity],
+      bufnr,
+      { lnum = diagnostic.range.start.line + 1 }
+    )
+  end
+end
+
+Diagnostic.buf_clear_saved_diagnostics = function(bufnr, diagnostic_ns, sign_ns)
```

Comment:
Should really decide on a naming patter for these or some obvious way to say that "buf_" means the first argument will not be the type, but something else. I liked the pattern I had going before this module